File: 2024/Deno/07.ts
const data = (await Deno.readTextFile("../inputs/07.txt")).trim().split("\n");

function allValid(arrNums: number[], res: number): number[] {
    if (arrNums.length === 1) return [arrNums[0]];
    const first = arrNums[0];
    return allValid(arrNums.slice(1), res).flatMap((num) => {
        const sum = first + num, product = first * num;
        return [sum, product].filter((x) => x <= res);
    });
}

function allValidTwo(arrNums: number[], res: number): number[] {
    if (arrNums.length === 1) return [arrNums[0]];
    const first = arrNums[0];
    const magnitude = 10 ** (Math.floor(Math.log10(first)) + 1);
    return allValidTwo(arrNums.slice(1), res).flatMap((num) => {
        const sum = first + num,
            product = first * num,
            concat = num * magnitude + first;
        return [sum, product, concat].filter((x) => x <= res);
    });
}

let [numValid, numValidTolerances] = [0, 0];

for (const line of data) {
    const [a, b] = line.split(":");
    const target = Number(a);
    const numbers = b.trim().split(" ").map(Number).reverse();

    if (allValid(numbers, target).includes(target)) numValid += target;
    else if (allValidTwo(numbers, target).includes(target)) {
        numValidTolerances += target;
    }
}

console.log(numValid);
console.log(numValidTolerances);

File: 2024/Python/07.py
data = open("../inputs/07.txt").read().splitlines()

num_valid = 0
num_valid_tolerances = 0


def all_valid(arr_nums, res):
    if len(arr_nums) == 1:
        return [arr_nums[0]]
    return [
        s
        for num in all_valid(arr_nums[1:], res)
        for s in (arr_nums[0] + num, arr_nums[0] * num)
        if s <= res
    ]


def all_valid_two(arr_nums, res):
    if len(arr_nums) == 1:
        return [arr_nums[0]]
    return [
        s
        for num in all_valid_two(arr_nums[1:], res)
        for s in (
            arr_nums[0] + num,
            arr_nums[0] * num,
            num * 10 ** len(str(arr_nums[0])) + arr_nums[0],
        )
        if s <= res
    ]


for line in data:
    a = int(line.split(":")[0])
    b = [*map(int, line.split(":")[1].strip().split())][::-1]
    if a in all_valid(b, a):
        num_valid += a
    elif a in all_valid_two(b, a):
        num_valid_tolerances += a

print(num_valid, num_valid_tolerances)
